opte-api/src/ndp.rs

```diff
@@ -0,0 +1,134 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2022 Oxide Computer Company
+
+//! Types for working with the IPv6 Neighbor Discovery Protocol
+
+use super::Ipv6Addr;
+use super::MacAddr;
+use core::fmt::{self, Debug, Display};
+
+/// A Neighbor Discovery Protocol Router Advertisement, generated in response to
+/// a Router Solicitation.
+#[derive(Clone, Copy, Debug)]
+pub struct RouterAdvertisement {
+    /// The expected MAC address of the guest whose Router Solicitations we
+    /// respond to.
+    pub src_mac: MacAddr,
+
+    // The source IPv6 address we expect the solicitation from, which is the
+    // EUI-64 transform of the expected MAC address.
+    src_ip: Ipv6Addr,
+
+    /// The MAC address advertised by the router.
+    pub mac: MacAddr,
+
+    // The IPv6 address advertised by the router, which is the EUI-64 transform
+    // of the source MAC address.
+    ip: Ipv6Addr,
+
+    /// Managed address configuration, indicating that the peer can use DHCPv6
+    /// to acquire an IPv6 address.
+    pub managed_cfg: bool,
+}
+
+impl RouterAdvertisement {
+    /// Create new `RouterAdvertisement`.
+    ///
+    /// The `src_mac` is the expected source MAC address a Router Solicitation
+    /// should come from. The expected source IPv6 address will be derived from
+    /// this, as a link-local address using the EUI-64 transform.
+    ///
+    /// `mac` is the destination MAC the Solitications are expected to be
+    /// delivered to, and the source MAC of the generated Advertisement. The
+    /// IPv6 address is also derived from this.
+    ///
+    /// `managed_cfg` is set to `true` to indicate that the host can get further
+    /// configuration from a DHCPv6 server running on the network.
+    pub fn new(src_mac: MacAddr, mac: MacAddr, managed_cfg: bool) -> Self {
+        let ip = Ipv6Addr::from_eui64(&mac);
+        let src_ip = Ipv6Addr::from_eui64(&src_mac);
```

Comment:
I've been thinking more about this. It seems like we should not expect that guests use the EUI-64 transform to generate their link-local addresses. @rcgoodfellow pointed out that some versions of Fedora and RHEL may be configured to do something else. It seems sufficient for us to respond to any inbound RS from the guest's MAC, with any link-local address in `fe80::/10`.

Reply:
RFC 4291 IP Version 6 Addressing Architecture, which originally specified that interface identifiers be constructed in a Modified EUI-64 format, was updated by RFC 7136 Significance of IPv6 Interface Identifiers. That RFC modifies the statements of RFC 4291 to specify that Modified EUI-64 construction is required only when an interface identifier is derived from a MAC address.

While the motivations for using an interface identifier other than one derived from a MAC address are completely vacuous for link-local addresses (the MAC address from which the interface identifier is derived is by definition in the packet already) - no specific provisions were made in the RFC 7136 for link-local addresses. Because of that, there can be no realistic expectation of any sort of deterministic structure for interface identifiers within link-local addresses, including the Modified EUI-64 that was originally specified.

Reply:
Thanks Ry. The summary then is we need to just use `fe80::/10`, and assume nothing else about the source IP address, yeah?

Reply:
Yup. I'm afraid so.

Reply:
Thanks for confirming. It's no big deal, @rzezeski has made it trivial to change that kinda thing in OPTE. I'll update this and re-push.